### PR TITLE
Add NES emulation runtime benchmark

### PR DESCRIPTION
**Pull request by koute**
This PR adds a new real-world runtime benchmark to the suite: [Pinky](https://github.com/koute/pinky), my NES emulator.

It uses a simple test ROM I've found [here](https://github.com/christopherpow/nes-test-roms/tree/97720008e51db15dd281a2a1e64d4c65cf1bca4c/nes15-1.0.0) and emulates its execution for a given number of frames (which can be freely adjusted to make the benchmark shorter/longer).

### CODE REVIEW

**Comment by Kobzol** [comment]
Great! I have wanted to include some kind of interpreter in the suite, an emulator should do nicely.

I wanted to avoid the git dependency, but then I noticed that we already have a few of them in compile benchmarks, so it's probably not such a big deal.

I wonder if in the benchmark loop, the framebuffer has to be written out? Couldn't we just benchmark the emulation of instructions itself?

**Comment by koute** [comment]
> I wonder if in the benchmark loop, the framebuffer has to be written out? Couldn't we just benchmark the emulation of instructions itself?

I added it because technically that is *the* output of the emulator, but practically speaking it probably doesn't make much difference so I can remove it if you want. (It's just a simple loop which goes over its internal palettized framebuffer, plugs each entry into a lookup table, and writes out the corresponding ABGR color.) 

There's also an audio callback that I've omitted, but I'm not sure whether the compiler is smart enough to delete the final parts of the audio processing pipeline if that's empty. Maybe we should `black_box` both the video and audio outputs just to be sure?

**Review by Kobzol — COMMENTED** [review]
I tried it without the framebuffer and the instruction count stays pretty much the same (~3.4 billion instructions). So I would just remove the framebuffer completely, and return `nes` from the benchmarking closure instead of the buffer. The returned value is `black_box`ed automatically by the benchmarking infrastructure.

Other than that one nit, looks good to me! I'll also let @nnethercote take a look.

**Comment by Kobzol** [comment]
Oh, one more thing, it would be nice if you included some short comment either to the top of the `main.rs` file, or above the benchmark function, to shortly describe what is the benchmark doing.

We should have a list of runtime benchmark descriptions in some README.md (same as for the compilation time benchmarks), but that's not prepared yet.

File: collector/runtime-benchmarks/nes/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                state: nes::State::new(),
+            };
+
+            nes.load_rom(ROM).unwrap();
+
+            let mut buffer: Vec<u32> = Vec::new();
+            buffer.resize(256 * 240, 0);
+
+            move || {
+                // Trigger the autosolve mechanism in the ROM.
```
Comment (nnethercote) [inline collector/runtime-benchmarks/nes/src/main.rs:33]: Maybe a little more explanation here? That comment is so brief it doesn't tell me much. And does it only apply to the `0..4` loop?

File: collector/runtime-benchmarks/nes/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                for _ in 0..4 {
+                    nes.execute_until_vblank().unwrap();
+                }
+
+                nes.press(nes::ControllerPort::First, nes::Button::Select);
+                nes.execute_until_vblank().unwrap();
+                nes.release(nes::ControllerPort::First, nes::Button::Select);
```
Comment (nnethercote) [inline collector/runtime-benchmarks/nes/src/main.rs:40]: Do these button presses get past a starting menu into the actual game, or something like that? Again, a little more explanation would be helpful.

**Review by nnethercote — APPROVED** [review]
I think this is a great choice for a benchmark.

**Comment by koute** [comment]
Sorry for the delay!

I simplified the benchmark and added more comments as requested.

**Review by Kobzol — APPROVED** [review]
Thanks, looks good to me!

I'll let Nick react to his remarks, but I think that this is good to go now.

**Review by nnethercote — APPROVED** [review]
Lovely, thank you!

**Pull request merged**